rails_event_store/lib/rails_event_store/async_handler_helpers.rb

```diff
@@ -2,21 +2,32 @@
 
 module RailsEventStore
   module AsyncHandler
-    def self.with_defaults
-      with
-    end
+    class << self
+      def with_defaults
+        with
+      end
 
-    def self.with(event_store: Rails.configuration.event_store, event_store_locator: nil, serializer: RubyEventStore::Serializers::YAML)
-      Module.new do
-        define_method :perform do |payload|
-          event_store = event_store_locator.call if event_store_locator
-          super(event_store.deserialize(serializer: serializer, **payload.symbolize_keys))
+      def with(event_store: Rails.configuration.event_store, event_store_locator: nil, serializer: RubyEventStore::Serializers::YAML)
+        Module.new do
+          define_method :perform do |payload|
+            event_store = event_store_locator.call if event_store_locator
+            record = AsyncHandler.event_from_payload(payload, event_store, serializer)
+            super(record)
+          end
         end
       end
-    end
 
-    def self.prepended(host_class)
-      host_class.prepend with_defaults
+      def prepended(host_class)
+        host_class.prepend with_defaults
+      end
+
+      # @return [Event] an event as deserialized from the payload
+      def event_from_payload(payload, event_store, serializer)
+        return payload if payload.kind_of?(RailsEventStore::Event)
```

Comment:
We don't require Event objects in RES to be of this kind. It's likely they are but we accept anything that has the same interface as `RES::Event`.